### ISSUE TITLE
Feature Request: Allow dashes in context keys

### ISSUE THREAD

**Issue by jimmiebfulton** [Keats/tera#647]
I have developed a code generator that leverages Tera (Archetect).  At the time I wrote Archetect, there were small changes I needed to make in order to leverage Tera the way that I wanted to, so I vendored Tera into the project.  At this time there are very few changes need to make Tera work for my project, and I would like to instead depend on Tera as a library, removing the Tera vendoring from the project.

This code generator is a lot like kickstart, but I've added quite a few sophisticated features to it.  One of the features is around "smart casing", where I supply a number of extra filters to Tera for reshaping user inputs into various casings.  I'm using 'heck' library to do this.  As an example, I can do this:

```yaml
script:
    - set:
        project:
          prompt: "Project Name:"

        suffix:
          prompt: "Project Suffx:"
          default: "Service"

        ProjectPrefix:
          value: "{{ project | pascal_case }}"

        projectPrefix:
          value: "{{ project | camel_case }}"

        project-prefix:
          value: "{{ project | train_case }}"
```

From here, I can use these variable names throughout a templated directory structure, using the case-shaped variables as needed throughout the project.  As an example:

```java
public class {{ ProjectPrefix }} {

    private {{ ProjectPrefix }} {{ projectPrefix}}Instance = new {{ ProjectPrefix }}();
}
```

Unfortunately, Tera does not allow for identifiers with a dash in the name (Ex: `project-prefix`), and this precludes me from allowing users to create variables that match the casing style as they would use it within the project.  I have added this in the `tera.pest` file in my vendored version of Tera, and have been using this for the past couple of years without issue.

If this were added to Tera, it would likely be a departure from Jinja2.  However, it does allow a bit more flexibility.  Also, because this is in the pest file, this couldn't just be turned on as an optional feature, so this would need to be something Tera is willing to allow.

Essentially, I've changed this in tera.pest:

```
all_chars = _{'a'..'z' | 'A'..'Z' | "_" | '0'..'9'}
```
to this:
```
all_chars = _{'a'..'z' | 'A'..'Z' | "_" | "-" | '0'..'9'}
```

Would you be willing to allow for dashes in Context identifiers?  If so, I'd be happy to put up a PR for this change.  If you would like to keep identifiers matching Jinja2 conventions, I'll need to keep Tera vendored in my project.

Thanks for the consideration.

**Comment by Keats** [comment]
> there were small changes I needed to make in order to leverage Tera the way that I wanted to

What are those?

As for the issue itself that can't work and you should probably not have that in your vendored version either. That introduces ambiguity in the syntax. Is `{{ a-b }}` a key from the context or `a - b` the math operation? Making that change would probably mean you can't have subtraction of variables which might work in some narrow context but not in general.

**Comment by jimmiebfulton** [comment]
Ah.  Excellent point.

At the time I did the original vendoring, Tera used Error Chain, which I wasn’t keen on adding to my project.  Also, Tera didn’t have a way to simply render a string with a one-off template using a Tera instance with all of the custom filters that needed to be applied.  Those issues have since been resolved.

The way I’m leveraging Tera is more specialized than generalized.  It’s specifically for generating templated files and directory structures.  The only math I’ve ever had to use was for addition (incrementing a server port + 1 to set the default for a prompted maintenance port), which is why I never noticed a broken subtraction.  I’m willing to live with requiring subtraction and other math operations to have spaces on either side as a trade-off for the much more commonly used (in my case) multiple alternately cased variables.  I arrived at this convention due to the fact that a “|” symbol is not allowed in file names on windows.  I also didn’t like having a bunch of “{{ project | train_case }}” strewn throughout the project.  Certainly, I could have variable names like “{{ projectNameDashed }}”, but just doesn’t flow nearly as well as having every variable shaped like the use case.  I’ll make adjustments to ensure subtraction is supported provided there are spaces around the operators.  Clearly, this doesn’t work for Tera.  Thanks for pointing out the obvious. ;-)
